### You're doing great picking this up.  Just keep working hard.  Here's some inspiration.

### *I hated every minute of training. But I said, “Don’t quit. Suffer now and live the rest of your life as a champion.”*
#### – Muhammad Ali

----

YOUR NAME HERE: Nick Hertzog

----

## NOTE:  THERE ARE DASHES SURROUNDING EVERY QUESTION. DO NOT DELETE THESE!  

They are important for grading.  I am trying a new technique.


-----

# Tables (DataFrames)

Tables/DataFrames are a fundamental object type for representing data sets. They can be viewed in two ways:
* a sequence of named columns that each describe a single aspect of all entries in a data set, or
* a sequence of rows that each contain all information about a single entry in a data set.

In order to use tables/DataFrames, import two packages: ```numpy``` as np   and also ```pandas``` as pd.

In [4]:
import pandas as pd
import numpy as np

Empty tables can be created using the `DataFrame()` function which lives in the ```pandas``` module. An empty table is useful because it can be extended to contain new rows and columns.

In [5]:
table = pd.DataFrame()
table

Empty DataFrame
Columns: []
Index: []

Now you try!

In [9]:
flowers = pd.DataFrame()
flowers

Empty DataFrame
Columns: []
Index: []

You can add a column by using hard bracet notation. 

```table_name["column_name"] = np.array([element1, element2, ..., last_element])```

Let's add a column called "Number_of_Petals" with the values 8, 34, 5

In [10]:
flowers["Number_of_Petals"] = np.array([8, 34, 5])
flowers

Number_of_Petals
0                 8
1                34
2                 5

# **- - - - - - - - - - - - - - - - - - - - - - - - - - -**
# **Question 1**

Now you try.  Add a column called "Name" that has the values: 'lotus', 'sunflower', 'rose'

In [11]:
flowers["Name"] = np.array(['lotus', 'sunflower', 'rose'])
flowers

Number_of_Petals       Name
0                 8      lotus
1                34  sunflower
2                 5       rose

Add a column called "Color" that has the values: 'pink', 'yellow', 'red'

In [12]:
flowers["Color"] = np.array(['pink', 'yellow', 'red'])
flowers

Number_of_Petals       Name   Color
0                 8      lotus    pink
1                34  sunflower  yellow
2                 5       rose     red

# **- - - - - - - - - - - - - - - - - - - - - - - - - - -**

Now, we could have created the above table all at once using a **Dictionary**.  In a dictionary, the symbol ```:``` means "please map".  So the following, says "Please map the column named `Color` to the strings `pink`, `yellow`, and `red`.

```"Color" : np.array(['pink', 'yellow', 'red'])```

Similar to how we use hard brackets `[ ]` to list items when making an array, we use curley braces `{ }` to list these mappings when creating dictionaries.

So we could have made the flowers table all at once like this:

In [ ]:
#Create the dictionary
flower_dict = { "Number_of_Petals" : np.array([8, 34, 5]),
             "Name" : np.array(['lotus', 'sunflower', 'rose']),
             "Color" : np.array(['pink', 'yellow', 'red'])
          }

flower_dict

Now a ```Dictionary``` type is different from a ```DataFrame``` type so we now just have to change the ```flower_dict``` to a ```DataFrame```.

In [ ]:
flowers = pd.DataFrame(flower_dict)
flowers

# **- - - - - - - - - - - - - - - - - - - - - - - - - - -**
# **Question 2**

Now you try.  Create a DataFrame called ```inventory``` that has these 2 columns.

* The Item_Name column should have "Teddy Bear", "Firetruck", "Soccer Ball"
* The Price column should have 18.99, 15.29,  24.00

Do so by first creating a Dictionary and then converting it to a DataFrame.

In [13]:
#create the mapping/dictionary
inventory_dict = {
    "Item_Name": np.array(['Teddy Bear', 'Firetruck', 'Soccer Ball']),
    "Price": np.array([18.99, 15.29, 24.00])
}

#convert it to a dataframe
inventory = pd.DataFrame(inventory_dict)
inventory

Item_Name  Price
0   Teddy Bear  18.99
1    Firetruck  15.29
2  Soccer Ball  24.00

# **- - - - - - - - - - - - - - - - - - - - - - - - - - -**

Creating tables in this way involves a lot of typing. If the data have already been entered somewhere, it is usually possible to use Python to read it into a table, instead of typing it all in cell by cell.

Often, tables are created from files that contain comma-separated values. Such files are called CSV files.

Below, we use the Table method `read_csv` to read a CSV file that contains salary data from the nba. The data are placed in a table named `nba`.

In [22]:
nba = pd.read_csv('nba_salaries.csv')
nba

PLAYER POSITION                TEAM  15-16 SALARY
0       Paul Millsap       PF       Atlanta Hawks     18.671659
1         Al Horford        C       Atlanta Hawks     12.000000
2     Tiago Splitter        C       Atlanta Hawks      9.756250
3        Jeff Teague       PG       Atlanta Hawks      8.000000
4        Kyle Korver       SG       Atlanta Hawks      5.746479
..               ...      ...                 ...           ...
412        Gary Neal       PG  Washington Wizards      2.139000
413     DeJuan Blair        C  Washington Wizards      2.000000
414  Kelly Oubre Jr.       SF  Washington Wizards      1.920240
415   Garrett Temple       SG  Washington Wizards      1.100602
416     Jarell Eddie       SG  Washington Wizards      0.561716

[417 rows x 4 columns]

To give you some  background: "The NBA is the highest paying professional sports league in the world," [reported CNN](http://edition.cnn.com/2015/12/04/sport/gallery/highest-paid-nba-players/) in March 2016. The table `nba` contains the salaries of all National Basketball Association players in 2015-2016.

Each row represents one player. The columns are:

| **Column Label**   | Description                                         |
|--------------------|-----------------------------------------------------|
| `PLAYER`           | Player's name                                       |
| `POSITION`         | Player's position on team                           |
| `TEAM`             | Team name                                           |
|`'15-'16 SALARY`    | Player's salary in 2015-2016, in millions of dollars|
 
The code for the positions is PG (Point Guard), SG (Shooting Guard), PF (Power Forward), SF (Small Forward), and C (Center). But what follows doesn't involve details about how basketball is played.

The first row shows that Paul Millsap, Power Forward for the Atlanta Hawks, had a salary of almost $\$18.7$ million in 2015-2016.

We will use this small table to demonstrate some useful DataFrame methods. 

### Size of the Table ###

The quickest way to see the number of rows and columns is to use the shape feature.  (Shape is not a method but a **feature/property** of DataFrames so there are no parentheses.)

In [23]:
#This should report (417, 4) 
#This means that the table has 417 rows and 4 columns
nba.shape

(417, 4)

Another way to get the size of the table is to:
* Analyze the length of the DataFrame for the number of rows
* Analyze the length of the columns feature

In [24]:
#Check out the length of the DF for the number of rows
len(nba)

417

In [25]:
#The columns FEATURE (so no parentheses) lists out the columns so we can take
#the length of that to find the number of columns.
len(nba.columns)

4

It is very typical to just want to see a few or the last few rows from the table.  We can use the ``head`` and ``tail'' functons to do just that.

In [26]:
#This shows the first 10 rows
nba.head(10)

PLAYER POSITION           TEAM  15-16 SALARY
0      Paul Millsap       PF  Atlanta Hawks     18.671659
1        Al Horford        C  Atlanta Hawks     12.000000
2    Tiago Splitter        C  Atlanta Hawks      9.756250
3       Jeff Teague       PG  Atlanta Hawks      8.000000
4       Kyle Korver       SG  Atlanta Hawks      5.746479
5   Thabo Sefolosha       SF  Atlanta Hawks      4.000000
6        Mike Scott       PF  Atlanta Hawks      3.333333
7     Kent Bazemore       SF  Atlanta Hawks      2.000000
8   Dennis Schroder       PG  Atlanta Hawks      1.763400
9  Tim Hardaway Jr.       SG  Atlanta Hawks      1.304520

In [27]:
#This shows the first 15 rows
nba.head(15)

PLAYER POSITION            TEAM  15-16 SALARY
0       Paul Millsap       PF   Atlanta Hawks     18.671659
1         Al Horford        C   Atlanta Hawks     12.000000
2     Tiago Splitter        C   Atlanta Hawks      9.756250
3        Jeff Teague       PG   Atlanta Hawks      8.000000
4        Kyle Korver       SG   Atlanta Hawks      5.746479
5    Thabo Sefolosha       SF   Atlanta Hawks      4.000000
6         Mike Scott       PF   Atlanta Hawks      3.333333
7      Kent Bazemore       SF   Atlanta Hawks      2.000000
8    Dennis Schroder       PG   Atlanta Hawks      1.763400
9   Tim Hardaway Jr.       SG   Atlanta Hawks      1.304520
10    Walter Tavares        C   Atlanta Hawks      1.000000
11  Jason Richardson       SG   Atlanta Hawks      0.947276
12   Lamar Patterson       SG   Atlanta Hawks      0.525093
13   Terran Petteway       SG   Atlanta Hawks      0.525093
14     Avery Bradley       PG  Boston Celtics      7.730337

In [28]:
#This shows the last 10 rows
nba.tail(10)

PLAYER POSITION                TEAM  15-16 SALARY
407   Kris Humphries       PF  Washington Wizards      4.440000
408     Jared Dudley       SF  Washington Wizards      4.375000
409    Alan Anderson       SG  Washington Wizards      4.000000
410      Drew Gooden       PF  Washington Wizards      3.300000
411   Ramon Sessions       PG  Washington Wizards      2.170465
412        Gary Neal       PG  Washington Wizards      2.139000
413     DeJuan Blair        C  Washington Wizards      2.000000
414  Kelly Oubre Jr.       SF  Washington Wizards      1.920240
415   Garrett Temple       SG  Washington Wizards      1.100602
416     Jarell Eddie       SG  Washington Wizards      0.561716

# **- - - - - - - - - - - - - - - - - - - - - - - - - - -**
# **Question 3**
Now you try.  In the following cells, you will read in the "countries.csv" file which has life expectancy and GDP data for several countries across a range of time.

In [15]:
#Read in the countries csv into a DataFrame called countries
countries = pd.read_csv('countries.csv')

In [18]:
#Find the number of nows
num_rows = len(countries)
num_columns = len(countries.columns)
print("Rows:", num_rows, " Columns:", num_columns)

Rows: 1704  Columns: 6


In [20]:
#Show the first 8 rows of the data frame
countries.head(8)

country continent  year  lifeExpectancy  population  gdpPerCapita
0  Afghanistan      Asia  1952          28.801     8425333    779.445314
1  Afghanistan      Asia  1957          30.332     9240934    820.853030
2  Afghanistan      Asia  1962          31.997    10267083    853.100710
3  Afghanistan      Asia  1967          34.020    11537966    836.197138
4  Afghanistan      Asia  1972          36.088    13079460    739.981106
5  Afghanistan      Asia  1977          38.438    14880372    786.113360
6  Afghanistan      Asia  1982          39.854    12881816    978.011439
7  Afghanistan      Asia  1987          40.822    13867957    852.395945

In [19]:
#Show the first 8 rows of the data frame
countries.head(8)

country continent  year  lifeExpectancy  population  gdpPerCapita
0  Afghanistan      Asia  1952          28.801     8425333    779.445314
1  Afghanistan      Asia  1957          30.332     9240934    820.853030
2  Afghanistan      Asia  1962          31.997    10267083    853.100710
3  Afghanistan      Asia  1967          34.020    11537966    836.197138
4  Afghanistan      Asia  1972          36.088    13079460    739.981106
5  Afghanistan      Asia  1977          38.438    14880372    786.113360
6  Afghanistan      Asia  1982          39.854    12881816    978.011439
7  Afghanistan      Asia  1987          40.822    13867957    852.395945

# **- - - - - - - - - - - - - - - - - - - - - - - - - - -**

### Column Labels 
We mentioned before that the `columns` feature of a DataFrame lists out teh actual column names.  With `nba` we don't gain much by this, but it can be very useful for tables that are so large that not all columns are visible on the screen.

In [29]:
nba.columns

Index(['PLAYER', 'POSITION', 'TEAM', '15-16 SALARY'], dtype='object')

We can change column labels using the `rename` method. This creates a new table and leaves `nba` unchanged.  

Notice that we use the ```dictionary``` syntax because we're really just mapping the old name to a new name.

In [30]:
nba.rename(columns={'15-16 SALARY':'SALARY'})

PLAYER POSITION                TEAM     SALARY
0       Paul Millsap       PF       Atlanta Hawks  18.671659
1         Al Horford        C       Atlanta Hawks  12.000000
2     Tiago Splitter        C       Atlanta Hawks   9.756250
3        Jeff Teague       PG       Atlanta Hawks   8.000000
4        Kyle Korver       SG       Atlanta Hawks   5.746479
..               ...      ...                 ...        ...
412        Gary Neal       PG  Washington Wizards   2.139000
413     DeJuan Blair        C  Washington Wizards   2.000000
414  Kelly Oubre Jr.       SF  Washington Wizards   1.920240
415   Garrett Temple       SG  Washington Wizards   1.100602
416     Jarell Eddie       SG  Washington Wizards   0.561716

[417 rows x 4 columns]

However, this method does not change the original table. 

In [31]:
nba

PLAYER POSITION                TEAM  15-16 SALARY
0       Paul Millsap       PF       Atlanta Hawks     18.671659
1         Al Horford        C       Atlanta Hawks     12.000000
2     Tiago Splitter        C       Atlanta Hawks      9.756250
3        Jeff Teague       PG       Atlanta Hawks      8.000000
4        Kyle Korver       SG       Atlanta Hawks      5.746479
..               ...      ...                 ...           ...
412        Gary Neal       PG  Washington Wizards      2.139000
413     DeJuan Blair        C  Washington Wizards      2.000000
414  Kelly Oubre Jr.       SF  Washington Wizards      1.920240
415   Garrett Temple       SG  Washington Wizards      1.100602
416     Jarell Eddie       SG  Washington Wizards      0.561716

[417 rows x 4 columns]

A common pattern is to assign the original name `nba` to the new table, so that all future uses of `nba` will refer to the relabeled table.

In [32]:
nba = nba.rename(columns={'15-16 SALARY':'SALARY'})
nba

PLAYER POSITION                TEAM     SALARY
0       Paul Millsap       PF       Atlanta Hawks  18.671659
1         Al Horford        C       Atlanta Hawks  12.000000
2     Tiago Splitter        C       Atlanta Hawks   9.756250
3        Jeff Teague       PG       Atlanta Hawks   8.000000
4        Kyle Korver       SG       Atlanta Hawks   5.746479
..               ...      ...                 ...        ...
412        Gary Neal       PG  Washington Wizards   2.139000
413     DeJuan Blair        C  Washington Wizards   2.000000
414  Kelly Oubre Jr.       SF  Washington Wizards   1.920240
415   Garrett Temple       SG  Washington Wizards   1.100602
416     Jarell Eddie       SG  Washington Wizards   0.561716

[417 rows x 4 columns]

Note that we can change multiple columns at once by adding additional mappings to our dictionary with commans, like follows.

In [33]:
nba = nba.rename(columns={'15-16 SALARY':'SALARY', 
                          'PLAYER':'PLAYER_NAME'})
nba

PLAYER_NAME POSITION                TEAM     SALARY
0       Paul Millsap       PF       Atlanta Hawks  18.671659
1         Al Horford        C       Atlanta Hawks  12.000000
2     Tiago Splitter        C       Atlanta Hawks   9.756250
3        Jeff Teague       PG       Atlanta Hawks   8.000000
4        Kyle Korver       SG       Atlanta Hawks   5.746479
..               ...      ...                 ...        ...
412        Gary Neal       PG  Washington Wizards   2.139000
413     DeJuan Blair        C  Washington Wizards   2.000000
414  Kelly Oubre Jr.       SF  Washington Wizards   1.920240
415   Garrett Temple       SG  Washington Wizards   1.100602
416     Jarell Eddie       SG  Washington Wizards   0.561716

[417 rows x 4 columns]

### Named Arguments

The `columns={'15-16 SALARY':'SALARY'}` portion of the call expression is called a *named argument*. When a function or method is called, each argument has both a position and a name. Both are evident from the help text of a function or method.   

To read the help text on ANY function, just type the function up to the parentheses and then include a '?', like below.  Notice that the columns argument is explained.  Look for the like that says ```columns : dict-like or function```.

In [34]:
nba.rename?

Signature:
nba.rename(
    mapper=None,
    index=None,
    columns=None,
    axis=None,
    copy=True,
    inplace=False,
    level=None,
    errors='ignore',
)
Docstring:
Alter axes labels.

Function / dict values must be unique (1-to-1). Labels not contained in
a dict / Series will be left as-is. Extra labels listed don't throw an
error.

See the :ref:`user guide <basics.rename>` for more.

Parameters
----------
mapper : dict-like or function
    Dict-like or functions transformations to apply to
    that axis' values. Use either ``mapper`` and ``axis`` to
    specify the axis to target with ``mapper``, or ``index`` and
    ``columns``.
index : dict-like or function
    Alternative to specifying axis (``mapper, axis=0``
    is equivalent to ``index=mapper``).
columns : dict-like or function
    Alternative to specifying axis (``mapper, axis=1``
    is equivalent to ``columns=mapper``).
axis : {0 or 'index', 1 or 'columns'}, default 0
    Axis to target with ``mapper``. Can be either

# **- - - - - - - - - - - - - - - - - - - - - - - - - - -**
# **Question 4**

Rename:

* the ```lifeExpectancy``` column in the countries DataFrame to ```life_exp```.

* the ```gdpPerCapita``` column in the countries DataFrame to ```gpd_per_capita```.

Be sure to update the countries DataFrame so that these changes are permanent.  

Challenge:  Can you change both column names in just 1 line?


In [49]:
#Your code here
countries = countries.rename(columns={'lifeExpectancy': 'life_exp', 'gdpPerCapita': 'gdp_per_capita'})
countries #We should see the change printed out

country continent  year  lifee_exp  population  gpd_per_capita
0     Afghanistan      Asia  1952     28.801     8425333      779.445314
1     Afghanistan      Asia  1957     30.332     9240934      820.853030
2     Afghanistan      Asia  1962     31.997    10267083      853.100710
3     Afghanistan      Asia  1967     34.020    11537966      836.197138
4     Afghanistan      Asia  1972     36.088    13079460      739.981106
...           ...       ...   ...        ...         ...             ...
1699     Zimbabwe    Africa  1987     62.351     9216418      706.157306
1700     Zimbabwe    Africa  1992     60.377    10704340      693.420786
1701     Zimbabwe    Africa  1997     46.809    11404948      792.449960
1702     Zimbabwe    Africa  2002     39.989    11926563      672.038623
1703     Zimbabwe    Africa  2007     43.487    12311143      469.709298

[1704 rows x 6 columns]

# **- - - - - - - - - - - - - - - - - - - - - - - - - - -**

### Accessing the Data in a Column ###
We can use a column's name to access the array of data in the column.

In [37]:
#Select the SALARY column
nba['SALARY']

0      18.671659
1      12.000000
2       9.756250
3       8.000000
4       5.746479
         ...    
412     2.139000
413     2.000000
414     1.920240
415     1.100602
416     0.561716
Name: SALARY, Length: 417, dtype: float64

Now, let's see why TYPE a column truly is.  Take a guess before running the next cell.

In [38]:
salary_col = nba['SALARY']
type(salary_col)

pandas.core.series.Series

I bet you figured you'd see "array" printed out.  So yup.  It turns out that **the columns of a DataFrane are technically of type series, not arrays.**  But we love working with arrays. so we can call the ```to_numpy``` function on a column/series to convert it to array.

In [39]:
salary_col = nba['SALARY'].to_numpy()
type(salary_col)

numpy.ndarray

Now we have an array!  Don't worry:  an ```ndarray``` is the same as a numpy array.  SO we can now use it to find the average salary.

In [40]:
#Remember- This is in millions of dollars.
salary_col.mean()

5.074814482014388

# **- - - - - - - - - - - - - - - - - - - - - - - - - - -**
# **Question 5**

Select the GDP PER CAPITA column of the countries DataFrame.  (Don't remember the exact name of the column?  Use contries.columns to reminder yourself!)

Then convert it to an array, and compute the max GDP PER CAPITA in hundreds of thousands of dollars.

HINT:  The answer will be 1.135231329.

In [57]:
countries.columns
gdp_as_array = countries['gpd_per_capita'].to_numpy()
mean_gdp_in_100_000s = (gdp_as_array / 100000).max()
mean_gdp_in_100_000s 

1.135231329

NOTICE:  In the previous question we found the max GDP of the data set, but we don't yet know how to retrieve the COUNTRY with the max GDP.  We'll know how next week!  For now, we could sort a table by different columns to determine this answer.

# **- - - - - - - - - - - - - - - - - - - - - - - - - - -**

### Sorting

Glance through about 20 rows if the ```nba``` table and you will see that the rows are in alphabetical order by team name. It's also possible to list the same rows in alphabetical order by player name using the `sort_values` method. The argument to `sort_values` is a column name.

In [54]:
nba.sort_values('PLAYER_NAME')

PLAYER_NAME POSITION                    TEAM     SALARY
68      Aaron Brooks       PG           Chicago Bulls   2.250000
291     Aaron Gordon       PF           Orlando Magic   4.171680
59    Aaron Harrison       SG       Charlotte Hornets   0.525093
235    Adreian Payne       PF  Minnesota Timberwolves   1.938840
1         Al Horford        C           Atlanta Hawks  12.000000
..               ...      ...                     ...        ...
95   Wilson Chandler       SF          Denver Nuggets  10.449438
233      Zach LaVine       PG  Minnesota Timberwolves   2.148360
181    Zach Randolph       PF       Memphis Grizzlies   9.638555
86     Zaza Pachulia        C        Dallas Mavericks   5.200000
26      Zoran Dragic       SG          Boston Celtics   1.706225

[417 rows x 4 columns]

To examine the players' salaries, it would be much more helpful if the data were ordered by salary.

Doing so arranges all the rows of the table in *increasing* order of salary, with the lowest salary appearing first. The output is a new table with the same columns as the original but with the rows rearranged.

In [55]:
nba.sort_values('SALARY')

PLAYER_NAME POSITION                 TEAM     SALARY
267  Thanasis Antetokounmpo       SF      New York Knicks   0.030888
327          Cory Jefferson       PF         Phoenix Suns   0.049709
326            Jordan McRae       SG         Phoenix Suns   0.049709
324         Orlando Johnson       SG         Phoenix Suns   0.055722
325            Phil Pressey       PG         Phoenix Suns   0.055722
..                      ...      ...                  ...        ...
131           Dwight Howard        C      Houston Rockets  22.359364
255         Carmelo Anthony       SF      New York Knicks  22.875000
72             LeBron James       SF  Cleveland Cavaliers  22.970500
29              Joe Johnson       SF        Brooklyn Nets  24.894863
169             Kobe Bryant       SF   Los Angeles Lakers  25.000000

[417 rows x 4 columns]

These figures are somewhat difficult to compare as some of these players changed teams during the season and received salaries from more than one team; only the salary from the last team appears in the table. Point Guard Phil Pressey, for example, moved from Philadelphia to Phoenix during the year, and might be moving yet again to the Golden State Warriors. 

The CNN report is about the other end of the salary scale – the players who are among the highest paid in the world. 

To order the rows of the table in *decreasing* order of salary, we must use `sort_values` with the option `ascending=False` and so we learn our second **named argument**.

In [56]:
nba.sort_values('SALARY', ascending=False)

PLAYER_NAME POSITION                 TEAM     SALARY
169             Kobe Bryant       SF   Los Angeles Lakers  25.000000
29              Joe Johnson       SF        Brooklyn Nets  24.894863
72             LeBron James       SF  Cleveland Cavaliers  22.970500
255         Carmelo Anthony       SF      New York Knicks  22.875000
131           Dwight Howard        C      Houston Rockets  22.359364
..                      ...      ...                  ...        ...
200         Elliot Williams       SG    Memphis Grizzlies   0.055722
324         Orlando Johnson       SG         Phoenix Suns   0.055722
327          Cory Jefferson       PF         Phoenix Suns   0.049709
326            Jordan McRae       SG         Phoenix Suns   0.049709
267  Thanasis Antetokounmpo       SF      New York Knicks   0.030888

[417 rows x 4 columns]

Kobe Bryant, may he rest in peace, was the highest paid at a salary of $\$25$ million in his final season with the Lakers (2015-2016). Notice that the MVP of that year, Stephen Curry, doesn't appear among the top 10. He is quite a bit further down the list, as we will see later.  FYI: Curry was first player in NBA history to be elected MVP by a unanimous vote.

# **- - - - - - - - - - - - - - - - - - - - - - - - - - -**
# **Question 6**

In the countries dataframe, you found that the max GDP PER CAPITA in hundreds of thousands of dollars is 1.135231329.  Using sorting to determine which country/year had this max.  Be sure to se the variables to the answers you discover.


In [63]:
countries = countries.sort_values('gpd_per_capita')
#Then set these variables.
country_with_max = countries['country'].values[0]
year_of_max = countries['year'].values[0]
print(country_with_max,'had the max population and this occurred in the year', 
      year_of_max, end='.')

Congo, Dem. Rep. had the max population and this occurred in the year 2002.

Now use sorting to find which country/year had the min population.


In [65]:
#Sort here
countries = countries.sort_values('population', ascending=True)
#Then set these variables.
country_min_pop = countries['country'].values[0]
year_min_pop = countries['year'].values[0]
print(country_min_pop,'had the minimum population and this occurred in the year', 
      year_min_pop, end='.')

Sao Tome and Principe had the minimum population and this occurred in the year 1952.

Hopefully, you found that the year of the min population was 1952.  

It would make sense that the year with the minimum population is the earliest year that appears in the data set.  Write code to determine what the minimum year in the table is.  Then set a variable to report your answer.

In [66]:
#Find the minimum year in the data set via code here.
min_year_in_data = countries['year'].min()
min_year_in_data

1952

# **- - - - - - - - - - - - - - - - - - - - - - - - - - -**

Let's do one last example to show you the benefits of using the help option available to you!

# **- - - - - - - - - - - - - - - - - - - - - - - - - - -**
# **Question 7**

All the functions/methods we use provide lots of arguments to help you get what you need.

Below use the help option (by using a ```?```) to read about the sort_values function.  You can start by typing either```nba.sort_values``` or ```pd.DataFrame.sort_values```, as the function is associated with the ```DataFrame``` type from the ```pandas``` package.

In [67]:
#Use the help option here
nba.sort_values?

Signature:
nba.sort_values(
    by,
    axis=0,
    ascending=True,
    inplace=False,
    kind='quicksort',
    na_position='last',
    ignore_index=False,
    key: Union[Callable[[ForwardRef('Series')], Union[ForwardRef('Series'), ~AnyArrayLike]], NoneType] = None,
)
Docstring:
Sort by the values along either axis.

Parameters
----------
        by : str or list of str
            Name or list of names to sort by.

            - if `axis` is 0 or `'index'` then `by` may contain index
              levels and/or column labels.
            - if `axis` is 1 or `'columns'` then `by` may contain column
              levels and/or index labels.

            .. versionchanged:: 0.23.0

               Allow specifying index or column level names.
axis : {0 or 'index', 1 or 'columns'}, default 0
     Axis to be sorted.
ascending : bool or list of bool, default True
     Sort ascending vs. descending. Specify list for multiple sort
     orders.  If this is a list of bools, must match the lengt

Notice the examples at the end.  The 3rd example shows you how to sort first by one column and then by another.

Following this example, sort the countries DataFrame first by year, then by continent, and then by the country columns.  

HINT- If you do this right, the index of the first few rows will be as follows: 24, 36, 120, 156, 192... 

In [72]:
countries.sort_values(by=['year', 'continent', 'country'])


country continent  year  lifee_exp  population  gpd_per_capita
24           Algeria    Africa  1952     43.077     9279525     2449.008185
36            Angola    Africa  1952     30.015     4232095     3520.610273
120            Benin    Africa  1952     38.223     1738315     1062.752200
156         Botswana    Africa  1952     47.622      442308      851.241141
192     Burkina Faso    Africa  1952     31.975     4469979      543.255241
...              ...       ...   ...        ...         ...             ...
1487     Switzerland    Europe  2007     81.701     7554661    37506.419070
1583          Turkey    Europe  2007     71.777    71158647     8458.276384
1607  United Kingdom    Europe  2007     79.425    60776238    33203.261280
71         Australia   Oceania  2007     81.235    20434176    34435.367440
1103     New Zealand   Oceania  2007     80.204     4115771    25185.009110

[1704 rows x 6 columns]

# **- - - - - - - - - - - - - - - - - - - - - - - - - - -**

### Analyzing Unique Entries

Let's say we want to see a list of all the team names in the nba data set.  We could start by selecting the "TEAM" column.

In [73]:
nba['TEAM']

0           Atlanta Hawks
1           Atlanta Hawks
2           Atlanta Hawks
3           Atlanta Hawks
4           Atlanta Hawks
              ...        
412    Washington Wizards
413    Washington Wizards
414    Washington Wizards
415    Washington Wizards
416    Washington Wizards
Name: TEAM, Length: 417, dtype: object

Why this is a good start, notice that every team is repeated many times.  We can use the ```unique``` method to request a list where all repeats have been eliminated.  So we say that a list with no repeated values is a list in which every value is **unique**.

In [74]:
unique_teams = nba["TEAM"].unique()

By combining this idea with the length function, we have an easy way to determine that there are 30 unique teams are in our data set.

In [75]:
len(unique_teams)

30

# **- - - - - - - - - - - - - - - - - - - - - - - - - - -**
# **Question 8**

Are all continents in the world represented in the countries data set?  Use the ```unique``` function combined with the ```length``` function to start your analysis.  Feel free to use other additional analyses if needed as well.  Then write your answer in the markdown area provided.  You should how many continents there are if you forgot. :)

In [77]:
#Your code here.  

#len(countries['continent'].unique())
countries['continent'].unique()

array(['Africa', 'Asia', 'Europe', 'Americas', 'Oceania'], dtype=object)

*Write your answer to the question here:  Are **all** continents represented?  Which are?  Which aren't? Is this sensible?*

# **Africa, Asia, Europe, Americas and Oceania are represented. I believe they are counting north and south america together. The only country then missing would be antartica which makes sense because officially, there are no permanent residents on antartica. People on antartica only stay for a few months to a year at a time on a scientific research base.**

### NOW SAVE AND SUBMIT!  USE THE CORRECT NAMING CONVENTION!

:)